src/casp/ml/data/distributed_anndata_sampler.py

```diff
@@ -0,0 +1,80 @@
+from typing import Iterator, Optional, TypeVar
+
+import torch
+import torch.distributed as dist
+from torch.utils.data import Dataset
+from torch.utils.data.distributed import DistributedSampler
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+class DistributedAnnCollectionSampler(DistributedSampler):
+    def __init__(
+        self,
+        dataset: Dataset,
+        shard_size: int,
+        seed: int = 0,
+        shuffle: bool = True,
+        num_replicas: Optional[int] = None,
+        rank: Optional[int] = None,
+    ) -> None:
+        if num_replicas is None:
+            if not dist.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            num_replicas = dist.get_world_size()
+        if rank is None:
+            if not dist.is_available():
+                raise RuntimeError("Requires distributed package to be available")
+            rank = dist.get_rank()
+        if rank >= num_replicas or rank < 0:
+            raise ValueError(
+                "Invalid rank {}, rank should be in the interval" " [0, {}]".format(rank, num_replicas - 1)
+            )
+
+        self.dataset = dataset
+        self.shard_size = shard_size
+        self.seed = seed
+        self.shuffle = shuffle
+        self.num_replicas = num_replicas
+        self.rank = rank
+
+        # TODO: several ways epoch are set, decide on one and use it
```

Comment:
Pytorch/Lightning can recreate the dataloaders at each epoch, in this is fine.  However, if you don't do that it can call the `set_epoch` method… the "problem" with that is all the work done here in the initializer that depends on the epoch should be moved into `iter()`.  

I think I prefer the latter since it works in either scenario, but would like to discuss